src/webgpu/api/validation/encoding/cmds/setImmediates.spec.ts

```diff
@@ -0,0 +1,209 @@
+export const description = `
+setImmediates validation tests.
+
+Test different encoder types (compute pass, render pass, render bundle):
+* Interpretation:
+  - Passing a TypedArray, the data offset and size are given in elements (not bytes).
+* Alignment:
+  - rangeOffset is not a multiple of 4 bytes.
+  - content size, converted to bytes, is not a multiple of 4 bytes.
+* Arithmetic overflow
+  - rangeOffset + contentSize is overflow
+* Bounds:
+  - dataOffset + contentSize (in bytes) exceeds the content data size.
+  - rangeOffset + contentSize (in bytes) exceeds the maxImmediateSize.
+`;
+
+import { makeTestGroup } from '../../../../../common/framework/test_group.js';
+import {
+  kTypedArrayBufferViewConstructors,
+  TypedArrayBufferViewConstructor,
+} from '../../../../../common/util/util.js';
+import { Float16Array } from '../../../../../external/petamoriken/float16/float16.js';
+import { AllFeaturesMaxLimitsGPUTest } from '../../../../gpu_test.js';
+import { kProgrammableEncoderTypes } from '../../../../util/command_buffer_maker.js';
+import { kMaxSafeMultipleOf8 } from '../../../../util/math.js';
+
+export const g = makeTestGroup(AllFeaturesMaxLimitsGPUTest);
+
+g.test('interpretation')
+  .desc('Tests that contentSize is interpreted as element size with TypedArray.')
+  .paramsSubcasesOnly(u =>
+    u //
+      .combine('encoderType', kProgrammableEncoderTypes)
+      .combine('success', [true, false])
+  )
+  .fn(t => {
+    const { encoderType, success } = t.params;
+
+    function runTest(arrayBufferType: TypedArrayBufferViewConstructor) {
+      const kMinAlignmentBytes = 4;
+      const elementSize = arrayBufferType.BYTES_PER_ELEMENT;
+      const maxImmediateSize = t.device.limits.maxImmediateSize; // using this limit for tests
+      const validDataSize = success
+        ? Math.floor((maxImmediateSize - kMinAlignmentBytes) / elementSize)
+        : maxImmediateSize - kMinAlignmentBytes;
+      const validOffset = success
+        ? Math.ceil(kMinAlignmentBytes / elementSize)
+        : kMinAlignmentBytes;
+
+      const { encoder, validateFinish } = t.createEncoder(encoderType);
+      const data = new arrayBufferType(maxImmediateSize);
+
+      encoder.setImmediates(/* rangeOffset */ 0, data, /* dataOffset */ validOffset, validDataSize);
+
+      validateFinish(success || elementSize === 1);
+    }
+
+    for (const arrayType of kTypedArrayBufferViewConstructors) {
+      if (arrayType === Float16Array) {
+        // Skip Float16Array since it is supplied by an external module, so there isn't an overload for it.
```

Comment:
Just a note: Float16Array is now supported by browsers natively, and we need to remove the external module, and remove any skips like this. https://github.com/gpuweb/cts/issues/4297